src/Accelerators/Accelerator.hpp

```diff
@@ -17,39 +17,53 @@
 #include "include/onnx-mlir/Compiler/OMCompilerTypes.h"
 #include "mlir/IR/BuiltinOps.h"
 #include "mlir/Pass/PassManager.h"
-#include <vector>
-
-extern bool InitAccelerators();
+#include "llvm/ADT/SmallVector.h"
 
 namespace onnx_mlir {
 namespace accel {
 
+// Note: this function is generated by cmake.
+extern bool initAccelerators();
+
 class Accelerator {
 public:
   /// Kinds of accelerators.
+  // clang-format off
   enum class Kind {
-#include "src/Accelerators/AcceleratorKinds.hpp"
+  // Note: AcceleratorKinds.inc is generated by cmake.
+  #include "src/Accelerators/AcceleratorKinds.inc"
   };
+  // clang-format on
 
   Accelerator(Kind kind) : kind(kind) {}
-  virtual ~Accelerator();
+  virtual ~Accelerator() = default;
 
   /// Getter for the kind of this accelerator.
   Kind getKind() const { return kind; }
 
-  static std::vector<Accelerator *> getAcceleratorList();
+  /// Returns the set of accelerators available.
```

Comment:
I find this slightly strange that the class `Accelerator` is the base class for each specialized accelerator (which is great), but also has a separate static list of accelerators. I would simply have a vector of accelerator (global symbol) where all valid accelerators are added. If one has the desire to ensure only the Accelerator subclasses are allowed to add to this class (I don't see why that would really be needed) then I would make a separate class `AcceleratorList` where only the `Accelerator` class (via a friend) can add a new accelerator.

Reply:
The reason this member function is static is because of the way we use it:
```
    onnx_mlir::accel::initAccelerators();
    for (auto *accel : onnx_mlir::accel::Accelerator::getAccelerators()) {
      if (!accel->isActive())
        continue;
      accel->getOrLoadDialects(context);
      accel->addPasses(module, pm, emissionTarget);
    }
```
In that code snippet we first initialize the accelerators availables. During initialization, each specific accelerator, adds itself to the list of accelerators maintained by the `Accelerator` base class. After initialization completes we then query the base class for that list/vector.  Now, because we cannot directly instantiate a Accelerator object (that class is virtual), we cannot make the accelerator list non-static.

Reply:
I know, I find just strange that the Accel is both a virtual class to be specialized by each custom accelerators as well as the holder of a list of accelerators. This functionality could be split in a separate accel list variable (or class, depending how fancy you want to be).